controllers/vcdmachine_controller.go

```diff
@@ -754,6 +764,113 @@ func (r *VCDMachineReconciler) reconcileNormal(ctx context.Context, cluster *clu
 	return ctrl.Result{}, nil
 }
 
+// getPrimaryNetwork returns the primary network based on vm.NetworkConnectionSection.PrimaryNetworkConnectionIndex
+// It is not possible to assume vm.NetworkConnectionSection.NetworkConnection[0] is the primary network when there are
+// multiple networks attached to the VM.
+func getPrimaryNetwork(vm *types.Vm) *types.NetworkConnection {
+	for _, network := range vm.NetworkConnectionSection.NetworkConnection {
+		if network.NetworkConnectionIndex == vm.NetworkConnectionSection.PrimaryNetworkConnectionIndex {
+			return network
+		}
+	}
+
+	return nil
+}
+
+// reconcileVMNetworks ensures that desired networks are attached to VMs
+// networks[0] refers the primary network
+func (r *VCDMachineReconciler) reconcileVMNetworks(vdcManager *vcdsdk.VdcManager, vApp *govcd.VApp, vm *govcd.VM, networks []string) error {
+	connections, err := vm.GetNetworkConnectionSection()
+	if err != nil {
+		return errors.Errorf("Failed to get attached networks to VM")
+	}
+
+	desiredConnectionArray := make([]*types.NetworkConnection, len(networks))
+
+	for index, ovdcNetwork := range networks {
+		err = ensureNetworkIsAttachedToVApp(vdcManager, vApp, ovdcNetwork)
+		if err != nil {
+			return errors.Errorf("Error ensuring network [%s] is attached to vApp", ovdcNetwork)
+		}
+
+		desiredConnectionArray[index] = getNetworkConnection(connections, ovdcNetwork)
+	}
+
+	if !containsTheSameElements(connections.NetworkConnection, desiredConnectionArray) {
+		connections.NetworkConnection = desiredConnectionArray
+		// update connection indexes for deterministic reconcilation
+		connections.PrimaryNetworkConnectionIndex = 0
+		for index, connection := range connections.NetworkConnection {
+			connection.NetworkConnectionIndex = index
+		}
+
+		err = vm.UpdateNetworkConnectionSection(connections)
+		if err != nil {
+			return errors.Errorf("failed to update networks of VM")
+		}
+	}
+
+	return nil
+}
+
+// containsTheSameElements checks all elements in the two array are the same regardless of order
```

Comment:
The time-complexity can be potentially improved using a small map with the key as the network name. That is a common pattern used and hence will be easier to read. This is not a big deal since the number of networks is small. Only the readability aspect is of interest, but this is also fine.